xmtp_mls/src/groups/mod.rs

```diff
@@ -233,36 +242,79 @@ where
         Self::create_from_welcome(client, provider, welcome)
     }
 
+    fn add_idempotency_key(encoded_msg: &[u8], idempotency_key: &str) -> PlaintextEnvelope {
+        PlaintextEnvelope {
+            content: Some(Content::V1(V1 {
+                content: encoded_msg.to_vec(),
+                idempotency_key: idempotency_key.into(),
+            })),
+        }
+    }
+
     pub async fn send_message(&self, message: &[u8]) -> Result<(), GroupError> {
         let conn = &mut self.client.store.conn()?;
 
-        let update_interval = Some(5_000_000); // 5 seconds in ns
+        let update_interval = Some(5_000_000); // 5 seconds in nanoseconds
         self.maybe_update_installations(conn, update_interval)
             .await?;
 
-        let intent_data: Vec<u8> = SendMessageIntentData::new(message.to_vec()).into();
+        let now = now_ns();
+        let plain_envelope = Self::add_idempotency_key(message, &now.to_string());
+        let mut encoded_envelope = vec![];
+        plain_envelope
+            .encode(&mut encoded_envelope)
+            .map_err(GroupError::EncodeError)?;
+
+        let intent_data: Vec<u8> = SendMessageIntentData::new(encoded_envelope).into();
         let intent =
             NewGroupIntent::new(IntentKind::SendMessage, self.group_id.clone(), intent_data);
         intent.store(conn)?;
 
+        // store this unpublished message locally before sending
+        let message_id = calculate_message_id(
+            &self.group_id,
+            message,
+            &self.client.account_address(),
+            &now.to_string(),
+        );
+        let group_message = StoredGroupMessage {
+            id: message_id,
+            group_id: self.group_id.clone(),
+            decrypted_message_bytes: message.to_vec(),
+            sent_at_ns: now,
```

Comment:
We are mixing client and server timestamps with this column, which might cause some confusion on clients where they are not aligned. But I'm okay with deferring that problem, this has the benefit that optimistic messages are automatically sorted for now. Perhaps we can leave a comment explaining this, in case we run into it down the line?

Reply:
Agree.  Its a side-effect of the main crux of this PR to store application messages.  I'll do a follow up PR to reconcile and unify these.